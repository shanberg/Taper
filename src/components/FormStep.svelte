<script lang="ts">
	import { createEventDispatcher } from 'svelte';
	import { isStepInvalid, isStepPlaceholder } from '../utils';
	export let steps: Step[];
	export let step: Step;
	export let index: number;

	const dispatch = createEventDispatcher();
	$: isOnlyRealStep = steps.length === 2;
	$: isPlaceholder = isStepPlaceholder(step);
	$: isInvalid = !isPlaceholder && isStepInvalid(step);
	$: isLastStep = index === steps.length - 1

	function handleClickDelete() {
		dispatch('removeStep', index);
	}
</script>

<div class="step" class:isInvalid class:isPlaceholder class:isLastStep>
	<div class="dose">
		<label for="dose-{index}">
			<input
				min={1}
				id="dose-{index}"
				step={step.dose > 5 ? 1 : 0.25}
				type="number"
				inputmode="decimal"
				pattern="[1-9]\d*"
				aria-label="dose-{index}"
				bind:value={step.dose}
				on:change={() => dispatch('change', step)}
			/>
		</label>
	</div>
	<div class="days">
		<label for="days-{index}">
			<input
				min={1}
				id="days-{index}"
				type="number"
				inputmode="decimal"
				pattern="[1-9]\d*"
				aria-label="days-{index}"
				bind:value={step.daysForDose}
				on:change={() => dispatch('change', step)}
			/>
		</label>
	</div>
	<button
		{...(isLastStep || isOnlyRealStep) ? { disabled: true } : {}}
		title="Remove this step"
		class="remove-btn"
		on:click={handleClickDelete}>Ã—</button
	>
</div>

<style>

.step {
	display: flex;
	flex: 0 0 12rem;
	gap: 1px;
}

	label {
		display: contents;
	}

	.dose, .days {
		flex: 1 1 100%;
	}

	input {
		width: 100%;
		height: 100%;
		border: none;
		outline: none;
	}

	.isPlaceholder {
		color: var(--color-text-muted);
		user-select: none;

		& input {
			background: transparent;
		}
	}

	.isInvalid input {
		background: var(--color-status-error-bg);
		color: var(--color-status-error);
	}

	.isInvalid:focus-within input {
		background: var(--color-status-error-background-muted);
	}

	.remove-btn {
		flex: 0 0 1.5rem;
		height: 100%;
		border: 0;
		margin-left: 1px;
		color: var(--color-text-muted);
		border-radius: var(--control-border-radius);
		transition: all var(--control-transition-duration) ease-in-out;
		background: transparent;
		cursor: pointer;
		opacity: 0.25;
		padding: 0;
		text-align: center;
	}

	.remove-btn:disabled {
		cursor: not-allowed;
	}

	.remove-btn:not(:disabled):where(:hover, :active) {
		opacity: 1;
		background: var(--color-status-error-background-muted);
	}

	.isLastStep .remove-btn {
		visibility: hidden;
	}

</style>
