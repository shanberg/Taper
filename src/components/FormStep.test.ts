import { describe, test, expect, vi, beforeEach } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import FormStep from './FormStep.svelte';

describe('FormStep', () => {
	let steps: Step[];
	let step: Step;
	let index: number;

	beforeEach(() => {
		steps = [
			{ dose: 1, daysForDose: 2 },
			{ dose: 3, daysForDose: 4 }
		];
		step = { dose: 1, daysForDose: 2 };
		index = 0;
	});

	test('renders correctly', () => {
		const { container } = render(FormStep, { steps, step, index });
		expect(container).toMatchSnapshot();
	});

	test('handles dose change', async () => {
		const { getByDisplayValue, component } = render(FormStep, { steps, step, index });
		const input = getByDisplayValue('1');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '2' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...step, dose: 2 }
			})
		);
	});

	test('handles daysForDose change', async () => {
		const { getByDisplayValue, component } = render(FormStep, { steps, step, index });
		const input = getByDisplayValue('2');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '3' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...step, daysForDose: 3 }
			})
		);
	});

	test('handles remove step', async () => {
		const { getByTitle, component } = render(FormStep, { steps, step, index });
		const button = getByTitle('Remove this step');

		const removeHandler = vi.fn();
		component.$on('removeStep', removeHandler);

		await fireEvent.click(button);

		expect(removeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: index
			})
		);
	});

	test('disables remove button when step is the only real step', () => {
		steps = [
			{ dose: 1, daysForDose: 1 },
			{ dose: 1, daysForDose: 1 }
		];
		const { getByTitle } = render(FormStep, { steps, step, index });
		const button = getByTitle('Remove this step');

		expect(button).toBeDisabled();
	});

	test('applies isPlaceholder class when step is a placeholder', () => {
		// isStepPlaceholder.mockReturnValue(true);
		const { container } = render(FormStep, {
			steps,
			step: { dose: 0, daysForDose: 0 },
			index
		});
		expect(container.querySelector('.step')).toHaveClass('isPlaceholder');
	});

	test('applies isInvalid class when step is invalid', () => {
		// isStepInvalid.mockReturnValue(true);
		const { container } = render(FormStep, {
			steps,
			step: { dose: -1, daysForDose: 0 },
			index
		});
		expect(container.querySelector('.step')).toHaveClass('isInvalid');
	});
});
