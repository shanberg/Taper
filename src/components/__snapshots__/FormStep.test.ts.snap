// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormStep > renders correctly 1`] = `
<div>
  <div
    class="step svelte-1p9l6v1"
  >
    <div
      class="dose svelte-1p9l6v1"
    >
      <label
        class="svelte-1p9l6v1"
        for="dose-0"
      >
        <input
          aria-label="dose-0"
          class="svelte-1p9l6v1"
          id="dose-0"
          inputmode="decimal"
          min="1"
          pattern="[1-9]\\d*"
          step="0.25"
          type="number"
        />
      </label>
    </div>
     
     
    <div
      class="duration svelte-1p9l6v1"
    >
      <label
        class="svelte-1p9l6v1"
        for="duration-0"
      >
        <input
          aria-label="duration-0"
          class="svelte-1p9l6v1"
          id="duration-0"
          inputmode="decimal"
          min="1"
          pattern="[1-9]\\d*"
          type="number"
        />
      </label>
    </div>
     
    <button
      class="remove-btn svelte-1p9l6v1"
      disabled=""
      title="Remove this step"
    >
      Ã—
    </button>
  </div>
</div>
`;
